ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -733,52 +739,95 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
     long totalCapacity = tgtLog.getCapacityInBytes();
     long writtenLastTime = 0;
     try (FileChannel fileChannel = Utils.openChannel(logSegmentToCopy.getView().getFirst(), false)) {
-      for (IndexEntry srcIndexEntry : srcIndexEntries) {
-        IndexValue srcValue = srcIndexEntry.getValue();
-        long usedCapacity = tgtIndex.getLogUsedCapacity();
-        if (isActive && (tgtLog.getCapacityInBytes() - usedCapacity >= srcValue.getSize())) {
-          fileChannel.position(srcValue.getOffset().getOffset());
-          Offset endOffsetOfLastMessage = tgtLog.getEndOffset();
-          // call into diskIOScheduler to make sure we can proceed (assuming it won't be 0).
-          diskIOScheduler.getSlice(DiskManager.CLEANUP_OPS_JOB_NAME, DiskManager.CLEANUP_OPS_JOB_NAME, writtenLastTime);
-          tgtLog.appendFrom(fileChannel, srcValue.getSize());
-          FileSpan fileSpan = tgtLog.getFileSpanForMessage(endOffsetOfLastMessage, srcValue.getSize());
-          if (srcValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
-            IndexValue putValue = tgtIndex.findKey(srcIndexEntry.getKey());
-            if (putValue != null) {
-              tgtIndex.markAsDeleted(srcIndexEntry.getKey(), fileSpan, srcValue.getOperationTimeInMs());
+      ListIterator<IndexEntry> srcIndexEntryIterator = srcIndexEntries.listIterator();
+      ByteBuffer bufferToUse = bundleReadBuffer;
+      List<IndexEntry> readyEntries = new ArrayList<>();
+      while (srcIndexEntryIterator.hasNext()) {
+        // reset startOffset to begin a new bundle read.
+        long startOffset = -1;
+        readyEntries.clear();
+        // get entries as many as possible and do a bunch read.
+        while (srcIndexEntryIterator.hasNext()) {
+          IndexEntry currentEntry = srcIndexEntryIterator.next();
+          if (startOffset == -1) {
+            startOffset = currentEntry.getValue().getOffset().getOffset();
+          }
+          readyEntries.add(currentEntry);
+          long readySize =
+              currentEntry.getValue().getOffset().getOffset() + currentEntry.getValue().getSize() - startOffset;
+          if (readySize > bundleReadBuffer.capacity()) {
+            // only have one in readyEntries and it exceeds bundleReadBuffer's capacity.
+            bufferToUse = ByteBuffer.allocate((int) readySize);
+            fileChannel.transferTo(startOffset, readySize,
+                Channels.newChannel(new ByteBufferOutputStream(bufferToUse)));
+            break;
+          }
+          long nextSize = Integer.MAX_VALUE;
+          if (srcIndexEntryIterator.hasNext()) {
+            // look ahead
+            IndexEntry nextEntry = srcIndexEntryIterator.next();
```

Comment:
I _think_ there is a bug here. I feel like this logic may be simpler if just iterate over index entries array using an index (or using it as a queue). Let's chat today and see what we can come up with.